### PR TITLE
Prefs [TASK][28546]: Apply MVP Structure

### PR DESCRIPTION
**Pull request by hieuna-441907**
## Related Tickets
- [#Task 28546: Apply MVP Structure](https://edu-redmine.sun-asterisk.vn/issues/28546)

## WHAT
- Apply MVP Structure
## HOW
- Apply MVP Structure
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/travelapp/ui/base/BasePresenter.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.sunasterisk.travelapp.ui.base
+
+interface BasePresenter {
+    fun start() {
+    }
+}
+
+
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/travelapp/ui/base/BasePresenter.kt:8]: remove blank

File: app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseRecyclerAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    private val items = mutableListOf<T>()
+
+    override fun getItemCount(): Int = items.size
+
+    override fun onBindViewHolder(viewHolder: V, position: Int) {
+        getItem(position)?.let { viewHolder.onBindData(it) }
+    }
+
+    protected fun getItem(position: Int): T? =
+        if (position in 0 until itemCount) items[position] else null
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseRecyclerAdapter.kt:17]: position liệu có bao giờ vượt quá range size không nhỉ mà ô phải check thế này?

File: app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseView.kt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.sunasterisk.travelapp.ui.base
+
+interface BaseView<T> {
+    var presenter: T
+}
+
+
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseView.kt:7]: remove blank

File: app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseViewHolder.kt [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    open fun onBindData(itemData: T) {
+        this.itemData = itemData
+    }
+
+    open fun onHandleItemClick(mainItem: T) {
+    }
+}
+
+
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseViewHolder.kt:24]: như trên

File: app/src/main/java/com/sunasterisk/travelapp/utils/ContextExt.kt [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    val msg = when (obj) {
+        is Int -> getString(obj)
+        is Exception -> obj.message.toString()
+        else -> obj.toString()
+    }
+    Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
+}
+
+
+
```
Comment (tuanpa-0908) [inline app/src/main/java/com/sunasterisk/travelapp/utils/ContextExt.kt:16]: rb

File: app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseActivity.kt [hunk 6]
```diff
@@ -0,0 +1,28 @@
+package com.sunasterisk.travelapp.ui.base
+
+import android.os.Bundle
+import androidx.appcompat.app.AppCompatActivity
+import androidx.fragment.app.Fragment
+
+abstract class BaseActivity : AppCompatActivity() {
+
+    protected abstract val layoutResource: Int
```
Comment (quockhanhng) [inline app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseActivity.kt:9]:
```suggestion
    @get:LayoutRes
    protected abstract val layoutResource: Int
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseFragment.kt [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    protected abstract val layoutResource: Int
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        super.onCreateView(inflater, container, savedInstanceState)
+        return inflater.inflate(layoutResource, container, false)
+    }
```
Comment (quockhanhng) [inline app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseFragment.kt:22]:
```suggestion
    ): View? = inflater.inflate(layoutResource, container, false)
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseActivity.kt [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+    protected abstract fun initComponents()
+
+    protected abstract fun initData()
+
+    protected fun openFragment(id: Int, fragment: Fragment, addToBackStack: Boolean = false) =
+        supportFragmentManager.beginTransaction().replace(id, fragment).apply {
+            if (addToBackStack) addToBackStack(null)
+        }.commit()
+}
```
Comment (quockhanhng) [inline app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseActivity.kt:27]: để ra extension được không anh

Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseActivity.kt:27]: ok, để đc em ạ

**Comment by hieuna-441907** [comment]
adf

**Pull request closed without merging**